### PR TITLE
add option no unchecked indexed access

### PR DESCRIPTION
**Pull request by ThomasJuster**

### CODE REVIEW

File: src/app/shared/helpers/null-undefined-predicates.ts [hunk 1]
```diff
@@ -10,3 +10,10 @@ export function isNotNull<T>(e: T|null): e is T {
 export function isNotUndefined<T>(e: T|undefined): e is T {
   return e !== undefined;
 }
+
+export const isDefined = isNotUndefined;
+
+export function ensureDefined<T>(value: T | undefined, message = 'expected value to be defined'): T {
```
Comment (smadbe) [inline src/app/shared/helpers/null-undefined-predicates.ts:16]: the error message should probably be "expected value to not to be undefined"

File: src/app/core/components/left-nav/left-nav.component.html [hunk 2]
```diff
@@ -33,18 +33,18 @@
...
   <span i18n>Unable to load the list.</span>&nbsp;
   <span (click)="retryError()" class="retry" i18n>Retry now</span>
 </p>
 
 <alg-left-nav-tree
-  *ngIf="dataSources[activeTabIndex].state.isReady"
-  [data]="dataSources[activeTabIndex].state.data"
+  *ngIf="dataSources[activeTabIndex]?.state?.isReady"
+  [data]="dataSources[activeTabIndex]?.state?.data"
   [elementType]="$any(['activity', 'skill', 'group'][activeTabIndex])"
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/core/components/left-nav/left-nav.component.html:48]: I agree, but angular template engine doesn't. I don't have an good idea of workaround for this one.
We could still add a property `activeDataSourceState` in the component and use it in the template, but I don't feel like the extra work is that useful. You tell me.

File: src/app/shared/helpers/null-undefined-predicates.ts (hunk 1 above)
Comment (ThomasJuster) [reply src/app/shared/helpers/null-undefined-predicates.ts]: Fixed

File: src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts [hunk 3]
```diff
@@ -93,20 +94,20 @@ export class GroupCompositionFilterComponent implements OnInit{
     const typeFilters = this.value.directChildren ? this.directChildrenTypeFilters : this.allDescendantsTypeFilters;
     if (index < 0 || index >= typeFilters.length) throw Error('invalid value for type filter');
     this.selectedTypeFilter = index;
-    this.value.type = typeFilters[index].value;
+    this.value.type = ensureDefined(typeFilters[index]).value;
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts:96]: Fixed

File: src/app/core/components/left-nav/left-nav.component.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/core/components/left-nav/left-nav.component.html:48]: Ok, let's leave it like that then.

**Review by smadbe — APPROVED** [review]
👍

**Pull request merged**